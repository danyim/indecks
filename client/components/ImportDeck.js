import React from 'react';
import { Link, browserHistory } from 'react-router';
import CSSTransitionGroup from 'react-addons-css-transition-group';

const ImportDeck = React.createClass({
  navigateToAddDeck(deckId) {
    browserHistory.push(`/add`);
  },

  render() {
    const deck = {
      title: '1',
      description: '2'
    };

    return (
      <section className="deck-view">
        <div>
        yo yo yo
          <h1>{deck.title}</h1>
          <p>{deck.description}</p>
        </div>
      </section>
    );
  }
})

export default ImportDeck;
