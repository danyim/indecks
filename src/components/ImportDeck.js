import React from 'react';
import { Link, browserHistory } from 'react-router';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as deckActions from '../action-creators/deck';
import CSSTransitionGroup from 'react-addons-css-transition-group';
import Dropzone from 'react-dropzone';
import Remarkable from 'remarkable';
import samples from '../data/samples'
import styles from '../styles/components/ImportDeck';

class ImportDeck extends React.Component {
  constructor(props) {
    super(props);
    this.handleDrop = this.handleDrop.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleLoadSample = this.handleLoadSample.bind(this);
  }

  generateRandomString(length = 8) {
    const chars = '0123456789abcdefABCDEFGHIJKLMNPQRSTUVWXYZ';
    let result = '';
    for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
    return result;
  }

  navigateToAddDeck(deckId) {
    browserHistory.push(`/add`);
  }

  handleDrop(files) {
    const file = files[0];
    const reader = new FileReader();
    reader.onload = e => {
      const result = e.target.result;
      const resultJson = JSON.parse(result);
      resultJson.id = this.generateRandomString(); // Generate a new ID regardless
      // Absolutely no validation of the JSON here...
      // We're trusting that the user is providing a indecks-produced deck json
      this.props.addDeck(resultJson);
      browserHistory.push(`/view/${resultJson.id}`);
      // TODO: Parse the contents of the card descriptions into Markdown
      // for(let card of resultJson.cards) {

      // }
      // const md = new Remarkable();
      // const markup = md.render(result);
      // this.setState({unprocessed: result, processed: markup});
      // console.log('complete!', result, markup, resultJson);
      // console.log('complete!', resultJson);
    }
    reader.readAsText(file);
  }

  handleSubmit(e) {
    e.preventDefault();
    const id = this.generateRandomString();
    this.props.addDeck({
      id: id,
      title: this.refs.title.value,
      description: this.refs.description.value,
      cards: []
    });

    browserHistory.push(`/view/${id}`);
  }

  handleLoadSample() {
    let sampleDeck;
    for(let d of samples) {
      sampleDeck = { ...d };
      sampleDeck.id = this.generateRandomString();
      this.props.addDeck(sampleDeck);
    }
    browserHistory.push(`/`);
  }

  render() {
    return (
      <section className={`${styles['deck-import']}`}>
        <div className={`${styles['grid-figure']}`}>
          <aside className={`${styles['figure-box']}`}>
            <div className={`${styles['symbol-container']}`}>
              <i className={`fa fa-list-alt fa-3 ${styles['symbol']}`}></i>
            </div>
          </aside>
          <form ref="commentForm" className="edit-form" onSubmit={this.handleSubmit}>
            <input type="text" className="large-input" name="title" ref="title" placeholder="Deck Title" maxLength="30" />
            <textarea type="text" name="description" ref="description" placeholder="Description" rows="3" />
            <Dropzone
              onDrop={this.handleDrop}
              className={`${styles['drop']}`}
              activeClassName={`${styles['drop-active']}`}
              rejectClassName={`${styles['drop-reject']}`}
              multiple={false}
              accept="application/json">
              <p>(Optional) Drop your deck JSON here</p>
            </Dropzone>
            <br /><br />
            <div>
              <button type="submit" className="button">Create</button>
            </div>
            <div>
              <button type="button" className="button" onClick={() => this.handleLoadSample()}>Load Sample Decks</button>
            </div>
          </form>
        </div>
      </section>
    );
  }
}

ImportDeck.defaultProps = {};
ImportDeck.propTypes = {
  addDeck: React.PropTypes.func.isRequired
};

const mapStateToProps = () => { return {} };
const mapDispatchToProps = (dispatch) => bindActionCreators(deckActions, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ImportDeck);

