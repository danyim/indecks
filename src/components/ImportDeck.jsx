import React from 'react';
import { browserHistory } from 'react-router';
import Dropzone from 'react-dropzone';
import samples from '../data/samples';
import styles from '../styles/components/ImportDeck';

const propTypes = {
  addDeck: React.PropTypes.func.isRequired,
  handleClose: React.PropTypes.func
};

const defaultProps = {
  handleClose: () => {}
};

class ImportDeck extends React.Component {
  constructor(props) {
    super(props);
    this.handleDrop = this.handleDrop.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleLoadSample = this.handleLoadSample.bind(this);
    this.renderCancelButton = this.renderCancelButton.bind(this);
  }

  generateRandomString(length = 8) {
    const chars = '0123456789abcdefABCDEFGHIJKLMNPQRSTUVWXYZ';
    let result = '';
    for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
    return result;
  }

  handleDrop(files) {
    const file = files[0];
    const reader = new FileReader();
    const addToDeck = inputDeck => {
      const deck = { ...inputDeck };
      deck.id = this.generateRandomString(); // Generate a new ID regardless
      // Absolutely no validation of the JSON here...
      // We're trusting that the user is providing a indecks-produced deck json
      this.props.addDeck(deck);
    };

    reader.onload = e => {
      const result = e.target.result;
      const resultJson = JSON.parse(result);
      if (Array.isArray(resultJson)) {
        // Trying to import multiple decks
        for (const d of resultJson) {
          addToDeck(d);
        }
        // Navigate to the deck grid view
        browserHistory.push('/');
      } else {
        // Add the single deck
        addToDeck(resultJson);
        // Close the modal
        this.props.handleClose();
        // Navigate to the newly imported deck
        browserHistory.push(`/view/${resultJson.id}`);
      }
      // resultJson.id = this.generateRandomString(); // Generate a new ID regardless
      // // Absolutely no validation of the JSON here...
      // // We're trusting that the user is providing a indecks-produced deck json
      // this.props.addDeck(resultJson);

      // TODO: Parse the contents of the card descriptions into Markdown
      // for(let card of resultJson.cards) {

      // }
      // const md = new Remarkable();
      // const markup = md.render(result);
      // this.setState({unprocessed: result, processed: markup});
      // console.log('complete!', result, markup, resultJson);
      // console.log('complete!', resultJson);
    };
    reader.readAsText(file);
  }

  handleSubmit(e) {
    e.preventDefault();

    if(this.refs.title.value.trim() === '') {
      return false;
    }

    const id = this.generateRandomString();
    this.props.addDeck({
      id,
      title: this.refs.title.value,
      description: this.refs.description.value,
      cards: []
    });

    // Close the modal
    this.props.handleClose();
    browserHistory.push(`/view/${id}`);
  }

  handleLoadSample() {
    let sampleDeck;
    for (const d of samples) {
      sampleDeck = { ...d };
      sampleDeck.id = this.generateRandomString();
      this.props.addDeck(sampleDeck);
    }

    // Close the modal
    this.props.handleClose();
    browserHistory.push('/');
  }

  renderCancelButton() {
    if (this.props.handleClose) {
      return (
        <div>
          <button
            type="button"
            className="button btn-delete"
            onClick={() => this.props.handleClose()}
          >
            Cancel
          </button>
        </div>
      );
    }
    return null;
  }

  render() {
    return (
      <section className={`${styles['deck-import']}`}>
        <div className={`${styles['grid-figure']}`}>
          <h2 className={`${styles['header']}`}>Add/Import Deck</h2>
          <form ref="commentForm" className="edit-form" onSubmit={this.handleSubmit}>
            <h4>Create a new deck</h4>
            <input type="text" className="large-input"
              name="title" ref="title"
              placeholder="Deck Title" maxLength="30" />
            <textarea type="text" name="description"
              ref="description" placeholder="Description" rows="3" />
            <div>
              <button type="submit" className="button">Create Deck</button>
            </div>
            <div className={`${styles['or-bar']}`}>
              <hr />
              <p>or</p>
            </div>
            <h4>Import an existing deck</h4>
            <Dropzone
              onDrop={this.handleDrop}
              className={`${styles['drop']}`}
              activeClassName={`${styles['drop-active']}`}
              rejectClassName={`${styles['drop-reject']}`}
              multiple={false}
              accept="application/json"
            >
              <p>Click here to import or drag and drop the deck JSON here</p>
            </Dropzone>
            <div className={`${styles['or-bar']}`}>
              <hr />
              <p>or</p>
            </div>
            <h4>Use samples</h4>
            <div>
              <button
                type="button" className="button"
                onClick={() => this.handleLoadSample()}
              >
                Load Sample Decks
              </button>
            </div>
            <br /><br />
            {this.renderCancelButton()}
          </form>
        </div>
      </section>
    );
  }
}

ImportDeck.propTypes = propTypes;
ImportDeck.defaultProps = defaultProps;

export default ImportDeck;
