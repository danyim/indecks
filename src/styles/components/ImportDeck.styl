@import '../_variables';

.deck-import
  max-width: pageWidth
  h4
    margin: 0
    margin-bottom: 1rem

.grid-figure
  flex-direction: column
  h2.header
    margin-top: 0
    padding-bottom: .3rem
    border-bottom: .6rem solid black
  form
    padding: 0

.drop
  width: 100%
  padding: 1rem 0
  margin: 1rem 0
  border: 1px dashed lighten(gray, 60%)
  text-align: center
  p
    color: lighten(gray, 60%)
.drop-active
  border: 1px solid green
.drop-reject
  border: 1px solid red

.or-bar
  hr
    border: 0
    border-bottom: 1px solid darken(lightgrey, 10%)
    padding: 0
    margin: 2rem 0
    width: 100%
    outline: 0
    animation-name: growWidth
    animation-duration: 3s
  p
    font-size: 1.2rem
    text-transform: uppercase
    position: absolute
    top: calc(2.0rem - .8rem)
    left: 50%
    transform: translate(-50%, 0%) !important
    color: darken(lightgrey, 10%)
    background-color: white
    margin: 0
    padding: 0 .6rem

@keyframes growWidth
    from
      width: 30%
    to
      width: 100%
