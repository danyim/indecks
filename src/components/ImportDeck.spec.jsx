import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { browserHistory } from 'react-router'
import ImportDeck from './ImportDeck'
// import samples from '../data/samples';

const defaultProps = {
  addDeck: () => {},
  maxDeckTitleLength: 160,
  maxDeckDescLength: 300
}

function setup(props = defaultProps) {
  const wrapper = shallow(<ImportDeck {...props} />)

  return {
    props,
    wrapper
  }
}

function setupFull(props = defaultProps) {
  const wrapper = mount(<ImportDeck {...props} />)

  return {
    props,
    wrapper
  }
}

describe('ImportDeck', () => {
  it('should render self and subcomponents', () => {
    const wrapper = shallow(<ImportDeck {...defaultProps} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should render a dropzone', () => {
    const { wrapper } = setupFull({
      ...defaultProps
    })

    expect(wrapper.find('Dropzone').exists()).toBe(true)
  })

  xit('should process the JSON when a file is provided in the Dropzone', () => {
    spyOn(browserHistory, 'push')
    spyOn(window, 'URL').and.returnValue({
      createObjectURL: () => {}
    })
    // spyOn(window, 'URL').and.returnValue(() => {});

    const handlerAdd = jest.fn()
    const { wrapper } = setupFull({
      ...defaultProps,
      addDeck: handlerAdd
    })

    const dropzone = wrapper.find('Dropzone')
    const files = [
      {
        name: 'test.json',
        size: 1111,
        type: 'application/json'
      }
    ]

    dropzone.simulate('drop', { dataTransfer: { files } })

    expect(handlerAdd.mock.calls.length).toBe(1)
    expect(browserHistory.push).toHaveBeenCalledWith('/')
  })

  it('should load 4 sample decks and then navigate home', () => {
    spyOn(browserHistory, 'push')
    const handlerAdd = jest.fn()
    const { wrapper } = setup({
      ...defaultProps,
      addDeck: handlerAdd
    })

    wrapper
      .find('button.button[children="Load Sample Decks"]')
      .simulate('click')
    expect(handlerAdd.mock.calls.length).toBe(4)
    expect(browserHistory.push).toHaveBeenCalledWith('/')
  })

  it('should take a title and description and call the addDeck prop', () => {
    const handlerAdd = jest.fn()
    const { wrapper } = setup({
      ...defaultProps,
      addDeck: handlerAdd
    })

    const title = wrapper.find('input[name="title"]')
    const description = wrapper.find('textarea[name="description"]')
    title.simulate('change', { target: { value: 'Test title' } })
    description.simulate('change', { target: { value: 'Test description' } })
    // console.log(wrapper.state());
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    })
    expect(handlerAdd.mock.calls.length).toBe(1)
  })

  it('should reject an empty title', () => {
    const handlerAdd = jest.fn()
    const { wrapper } = setup({
      ...defaultProps,
      addDeck: handlerAdd
    })

    const title = wrapper.find('input[name="title"]')
    title.simulate('change', { target: { value: '' } })

    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    })
    expect(handlerAdd.mock.calls.length).toBe(0)
  })

  it('should reject a title greater than maxDeckTitleLength', () => {
    const handlerAdd = jest.fn()
    const { props, wrapper } = setup({
      ...defaultProps,
      addDeck: handlerAdd
    })

    const longTitle = 'a'.repeat(props.maxDeckTitleLength + 1)

    const title = wrapper.find('input[name="title"]')
    const description = wrapper.find('textarea[name="description"]')
    title.simulate('change', { target: { value: longTitle } })
    description.simulate('change', { target: { value: '' } })

    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    })

    expect(handlerAdd.mock.calls.length).toBe(0)
  })

  it('should reject a description greater than maxDeckDescLength', () => {
    const handlerAdd = jest.fn()
    const { props, wrapper } = setup({
      ...defaultProps,
      addDeck: handlerAdd
    })

    const longDesc = 'a'.repeat(props.maxDeckDescLength + 1)

    const title = wrapper.find('input[name="title"]')
    const description = wrapper.find('textarea[name="description"]')
    title.simulate('change', { target: { value: 'Test' } })
    description.simulate('change', { target: { value: longDesc } })

    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    })

    expect(handlerAdd.mock.calls.length).toBe(0)
  })
})
