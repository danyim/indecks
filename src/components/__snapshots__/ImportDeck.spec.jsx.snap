// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportDeck should render self and subcomponents 1`] = `
<section
  className="deck-import"
>
  <div
    className="grid-figure"
  >
    <h2
      className="header"
    >
      Add/Import Deck
    </h2>
    <form
      className="edit-form"
      onSubmit={[Function]}
    >
      <h4>
        Create a new deck
      </h4>
      <input
        className="large-input"
        maxLength="30"
        name="title"
        onChange={[Function]}
        placeholder="Deck Title"
        type="text"
      />
      <textarea
        name="description"
        onChange={[Function]}
        placeholder="Description"
        rows="3"
        type="text"
      />
      <div>
        <button
          className="button"
          type="submit"
        >
          Create Deck
        </button>
      </div>
      <div
        className="or-bar"
      >
        <hr />
        <p>
          or
        </p>
      </div>
      <h4>
        Import an existing deck collection or deck
      </h4>
      <Dropzone
        accept="application/json"
        activeClassName="drop-active"
        className="drop"
        disableClick={false}
        disablePreview={false}
        maxSize={Infinity}
        minSize={0}
        multiple={false}
        onDrop={[Function]}
        preventDropOnDocument={true}
        rejectClassName="drop-reject"
      >
        <p>
          Click here to import or drag and drop the deck JSON here
        </p>
      </Dropzone>
      <div
        className="or-bar"
      >
        <hr />
        <p>
          or
        </p>
      </div>
      <h4>
        Use samples
      </h4>
      <div>
        <button
          className="button"
          onClick={[Function]}
          type="button"
        >
          Load Sample Decks
        </button>
      </div>
    </form>
  </div>
</section>
`;
